### ISSUE TITLE
Main function shouldn't wait for its workers to terminate

### ISSUE THREAD

**Issue by MaryamZi** [ballerina-platform/ballerina-lang#28536]
**Description:**
Right now, when the main function has worker declarations, the main function execution is considered to be complete only once the workers terminate. 

But the spec does not require this.

> The termination of a function is independent of the termination of its named workers. The termination of a named worker does not automatically result in the termination of its function. When a default worker terminates causing the function to terminate, the function does not automatically wait for the termination of its named workers.

Related slack discussion: https://ballerina-platform.slack.com/archives/CAXS256RX/p1612334915063100

**Steps to reproduce:**
```ballerina
import ballerina/io;
import ballerina/lang.runtime;

public function main() {
    io:println("Main - Start");
    worker w {
        int i = 0;
        while i < 3 {
            runtime:sleep(1);
            io:println("Worker iteration ", i);
            i += 1;
        }
    }
    io:println("Main - Done");
}  
```

Output
```cmd
$ bal run test.bal 

Compiling source
        test.bal

Running executable

Main - Start
Main - Done
Worker iteration 0
Worker iteration 1
Worker iteration 2
```

**Affected Versions:**
slalpha1

**Comment by jclark** [comment]
It would be good if graceful stop could deal with these workers. I will create a spec issue. My suggestion for now would be to require any named workers created by main or a module init function to be waited for.

**Comment by jclark** [comment]
If

- main calls a function `foo`
- `foo` has a named worker `bar`
- `foo` returns before `bar` terminates

does `main` wait for `bar`?

As the spec is set up currently, `main` is nothing special: it is called by `init` in an compiler-generated module.

**Comment by MaryamZi** [comment]
It seems like the initialization phase which calls the `init`/`main` function at the end waits for any workers started by any function (via the generated `init`/`main`) to complete (rather than the `main` waiting for its workers to complete like I mentioned initially).

> If
> 
> * main calls a function `foo`
> * `foo` has a named worker `bar`
> * `foo` returns before `bar` terminates
> 
> does `main` wait for `bar`?

So in such a scenario, if we only have a `main` function, the program will not terminate until `bar` is done (even though the `main` function completes the program does not terminate until the workers are done).

And when there are listeners, the listening phase does not begin until the workers started in the `init`/`main` function or in other functions called by them terminate.

Looping in @warunalakshitha and @manuranga too.

**Comment by warunalakshitha** [comment]
Yes. Actually, the main function also works the same as other functions. But the issue was the way we schedule main and start( which actually start the listener) which is done for two steps. We schedule the main and the scheduler will wait for end of main + any workers created by main or any other functions. After that, we schedule the start. So I think if we schedule the init, main and start as one function I think the above spec deviation can be fixed. But need to check on panic cases too.

**Comment by warunalakshitha** [comment]
We are already working on https://github.com/ballerina-platform/ballerina-lang/issues/27338 basically on compiler-generated init module. I hope with that this behavior will be fixed too.

**Comment by jclark** [comment]
Before you spend time fixing things, we should figure out what the right thing to do is. I don’t think the language team (at least me) have given thought to what should happen in this case.

When we have event streams, we may want to start strands to process unbounded streams generated by services (like `forever` used to do). One way to get these started would be in init: obviously in that case, we don’t want services to wait for those strands. But we would want a graceful stop for wait for those strands to be finished.

**Comment by pubudu91** [comment]
Ran in to this issue while trying to implement the concurrency example in https://en.wikipedia.org/wiki/Go_(programming_language)#Examples

```ballerina
import ballerina/io;
import ballerina/lang.runtime;

public function main() {
    @strand {thread: "any"}
    worker StdInReader returns string {
        string word = io:readln("Type a word, then hit enter: ");
        return word;
    }

    @strand {thread: "any"}
    worker Timer {
        runtime:sleep(5);
    }

    string? result = wait StdInReader | Timer;

    if result is string {
        io:println("Received: ", result);
    } else {
        io:println("Timed out");
    }
}
```
The expectation is for the program to exit once it times out. But this hangs after the time out till the `StdInReader` also completes.

**Comment by Nadeeshan96** [comment]
> Ran in to this issue while trying to implement the concurrency example in https://en.wikipedia.org/wiki/Go_(programming_language)#Examples

Created a [separate issue](https://github.com/ballerina-platform/ballerina-standard-library/issues/4663) for this. This need to be fixed from the module-ballerina-io. When the `io:readln` is called, it blocks the thread till it receives an input. Hence, we need to run the line read function asynchronously with a callback. This can be fixed with https://github.com/ballerina-platform/module-ballerina-io/pull/1054.

**Comment by Nadeeshan96** [comment]
This issue is fixed in Ballerina 2201.7.0 (Swan Lake Update 7). The first code snippet gives the output
```
Main - Start
Main - Done
```
and we can see the Ballerina program ends when the `main` function ends, since there is no explicit `wait` for the named workers in the default worker. Hence closing the issue.
